sklearn/ensemble/_forest.py

```diff
@@ -118,26 +123,47 @@ def _get_n_samples_bootstrap(n_samples, max_samples):
         return max_samples
 
     if isinstance(max_samples, Real):
-        return max(round(n_samples * max_samples), 1)
+        if sample_weight is None:
+            return max(int(max_samples * n_samples), 1)
+        else:
+            sw_sum = np.sum(sample_weight)
+            if sw_sum <= 1:
+                raise ValueError(
+                    f"The total sum of sample weights is {sw_sum}, which prevents "
+                    f"resampling with a fractional value for {max_samples=}. Either"
+                    " pass max_samples as an integer or use a larger sample_weight."
+                )
```

Comment:
I would swap the order of the possible suggestions, as only the first suggestion guarantees the weighting/repetition semantics.


```suggestion
                    f"The total sum of sample weights is {sw_sum}, which prevents "
                    f"resampling with a fractional value for {max_samples=}. Either"
                    " use a larger sample_weight or pass max_samples as an integer."
                )
```